lib/async_hooks.js

```diff
@@ -0,0 +1,486 @@
+    if (init && typeof init !== 'function')
+      throw new TypeError('init must be a function');
+    if (before && typeof before !== 'function')
+      throw new TypeError('before must be a function');
+    if (after && typeof after !== 'function')
+      throw new TypeError('after must be a function');
+    if (destroy && typeof destroy !== 'function')
+      throw new TypeError('destroy must be a function');
+
+    this[init_symbol] = init;
+    this[before_symbol] = before;
+    this[after_symbol] = after;
+    this[destroy_symbol] = destroy;
+  }
+
+  enable() {
+    // Use local reference in case the pointer to the array needs to change.
+    var hooks_array = active_hooks_array;
+    var hook_fields = async_hook_fields;
+
+    // If this hook is being enabled while in the middle of processing the
+    // array of currently active hooks then duplicate the current set of
+    // active hooks and store this there. This shouldn't fire until the next
+    // time hooks are processed.
+    if (processing_hooks) {
+      if (tmp_active_hooks_array === null)
+        storeActiveHooks();
+      hooks_array = tmp_active_hooks_array;
+      hook_fields = tmp_async_hook_fields;
+    }
+
+    // Each hook is only allowed to be added once.
+    if (hooks_array.includes(this))
+      return;
+
+    // createHook() has already enforced that the callbacks are all functions,
+    // so here simply increment the count of whether each callbacks exists or
+    // not.
+    hook_fields[kInit] += +!!this[init_symbol];
+    hook_fields[kBefore] += +!!this[before_symbol];
+    hook_fields[kAfter] += +!!this[after_symbol];
+    hook_fields[kDestroy] += +!!this[destroy_symbol];
+    hooks_array.push(this);
+    return this;
+  }
+
+  disable() {
+    // Use local reference in case the pointer to the array needs to change.
+    var hooks_array = active_hooks_array;
+    var hook_fields = async_hook_fields;
+
+    if (processing_hooks) {
+      if (tmp_active_hooks_array === null)
+        storeActiveHooks();
+      hooks_array = tmp_active_hooks_array;
+      hook_fields = tmp_async_hook_fields;
+    }
+
+    const index = hooks_array.indexOf(this);
+    if (index === -1)
+      return;
+
+    hook_fields[kInit] -= +!!this[init_symbol];
+    hook_fields[kBefore] -= +!!this[before_symbol];
+    hook_fields[kAfter] -= +!!this[after_symbol];
+    hook_fields[kDestroy] -= +!!this[destroy_symbol];
+    hooks_array.splice(index, 1);
+    return this;
+  }
+}
+
+
+function storeActiveHooks() {
+  tmp_active_hooks_array = active_hooks_array.slice();
+  // Don't want to make the assumption that kInit to kDestroy are indexes 0 to
+  // 4. So do this the long way.
+  tmp_async_hook_fields = [];
+  tmp_async_hook_fields[kInit] = async_hook_fields[kInit];
+  tmp_async_hook_fields[kBefore] = async_hook_fields[kBefore];
+  tmp_async_hook_fields[kAfter] = async_hook_fields[kAfter];
+  tmp_async_hook_fields[kDestroy] = async_hook_fields[kDestroy];
```

Comment:
Reason I don't `slice()` it here is because I don't need a full `Uint32Array()` (which takes longer to create).

I don't in `restoreTmpHooks()` b/c the `uint32_t*` that the `async_hook_fields` points to is stored on `Environment::AsyncHooks::fields_` and used to track the number of callbacks there are for a given hook across C++ and JS (which is the reason I make it a read-only property of `process.binding('async_wrap')`)